sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +291,111 @@ def reduce(self, word, exclude=None):
+
+        # Compute the proper prefixes for every rule.
+        for r in self.rules:
+            self.proper_prefixes[r] = []
+            letter_word_array = [s for s in r.letter_form_elm]
+            for i in range (1, len(letter_word_array)):
+                letter_word_array[i] = letter_word_array[i-1]*letter_word_array[i]
+            self.proper_prefixes[r] = letter_word_array
+
+        # Create the states in the automaton.
+        # The left-hand side of the rules are the dead states.
+        # The proper left-hand side of the rules are the accept states.
+
+        fsm = StateMachine('fsm')
+        fsm.add_state('start', is_start=True)
+
+        # Add dead states.
+        for rule in self.left_hand_rules:
+            if not rule in fsm.state_names:
+                fsm.add_state(rule, is_dead=True)
+
+        # Add accept states.
+        for r in self.rules:
+            prop_prefix = self.proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in fsm.state_names:
+                    fsm.add_state(elem, is_accept=True)
+
+        # Add transitions for every state
+        for state in fsm.states:
+            current_state_name = state.name
+            current_state_type = state.state_type
+            if current_state_type == "start":
+                for letter in self.automaton_alphabets:
+                    if letter in fsm.state_names:
+                        state.add_transition(letter, letter)
+                    else:
+                        state.add_transition(letter, current_state_name)
+            elif current_state_type == "accept":
+                for letter in self.automaton_alphabets:
+                    next = current_state_name*letter
+                    len_next_word = len(next)
+                    while True:
+                        if len(next) <= 1:
+                            if next in fsm.state_names:
+                                state.add_transition(letter, next)
+                            else:
+                                state.add_transition(letter, 'start')
+                            break
+                        else:
+                            if next in fsm.state_names:
+                                state.add_transition(letter, next)
+                                break
+                            next = next.subword(1, len(next))
+        return fsm
+
+    def reduce_using_automaton(self, word):
+        '''
+        The method for word reduction using automaton is mentioned in the section 13.1.3 of the Handook.
+        All the elements of the automaton are stored in an array and are given as the input to the automaton.
+        If the automaton reaches a dead state that subword is replaced and the automaton is run from the beginning.
+        This is repeated until the word reached the end and the automaton stays in the accept state.
+
+        '''
+        fsm = self.construct_automaton()
```

Comment:
Yeah, I was thinking of making `fsm` an attribute of the class. And instead of constructing when the `reduce _word` can we just construct/modify it once a new rule is added?

Reply:
Modifying it will be faster I expect but yeah, if you do modify it at any point, it would need to be done in functions that add new rules. Are you doing anything about rules being removed? This happens in `make_confluent` while new rules are being added.

Reply:
I think modifying might increase the complexity. 'Coz whenever a new rule is added we have to check for the redundant once and this might as well involve a few states to be deleted also the transitions need to be verified again at every existing state (to check if the there is a transition to the newly added state). 

So, instead of this, I think it would be better to dump the previous object instance and create a new one whenever the states are added/removed. And I think efficiency for both would be the same.

Reply:
Fair enough. In that case yes, definitely best to keep it immutable and just construct a new one when needed.